### DISCUSSION TITLE
Strain energy density calculation for hyperelastic material

### DISCUSSION THREAD

**Discussion by ada-ayy** [idaholab/moose#24107] [Q&A Modules: Solid mechanics] [answered]
Hello!
I am trying to model a composite consisting of a hyperelastic component and an elastic component. ​I would like to obtain the strain energy density of this whole system for further applications in PHASE FIELD modeling.

Is there any convenient way to obtain the strain energy density of a hyperelastic material? Or should I write a code to do this calculation?

I have tried to use **StrainEnergyDensity** material, but face the problem of stress definition in the hyperelastic region.
I use **ComputeNeoHookeanStress** to model holistic behavior. The following is some of my input:

> [Kernels]
>     [./sdx]
>       type = TotalLagrangianStressDivergence
>       variable = disp_x
>       component = 0
>       block = 'poly_QUAD4'
>     [../]
>     [sdy]
>       type = TotalLagrangianStressDivergence
>       variable = disp_y
>       component = 1
>       block = 'poly_QUAD4'
>     []
>     [./TensorMechanics]
>       displacements = 'disp_x disp_y'
>       block = 'fiber_QUAD4'
>     [../]
> []
> 
> 
> [Materials]
>  inactive = 'strain_energy_density'
>    [./compute_stress_A]
>      # type = ComputeStVenantKirchhoffStress
>       type = ComputeNeoHookeanStress
>       lambda = 4000.0
>       mu = 6700.0
>       block = 'poly_QUAD4'
>    [../]
>    
>    [./compute_stress_B]
>       type = ComputeLinearElasticStress
>       block = 'fiber_QUAD4'
>    [../]
>     
>    [./elasticity_tensor_B]
>     type = ComputeElasticityTensor
>     block = 'fiber_QUAD4'
>     fill_method = symmetric9
>     C_ijkl = '1.1e6 1e5 0 1e6 0 1e6 .5e6 .2e6 .5e6'
>   [../]
>    [./compute_strain]
>       type = ComputeLagrangianStrain
>    [../]
> 
>    [./strain_energy_density]
>       type = StrainEnergyDensity
>    [../]
> []

**Comment by hugary1995** [comment, marked as answer]
There is no one universal expression for the strain energy density function in a hyperelastic model, unfortunately. Parsing tensorial quantities is not supported yet in MOOSE. You will have to write some C++ to create a Material object to compute the strain energy density.

**Reply by ada-ayy** [reply]
Thanks for your reply! I will try
